### DISCUSSION TITLE
Load HTML string

### DISCUSSION THREAD

**Discussion by stychu** [cypress-io/cypress#14197] [Component Testing] [answered]
Is it possible to load the HTML string I cypress??

I need a way to assert some element existence when doing API calls.

cy.request('/').its('body') <- returns me a HTML string as a response
Can I then somehow load it in the cypress to make operations like on normal page load?

If I do directly
```
cy.request('/').its('body').should('contain', somestring)
```
It does the job but unfortunately, this loads the whole HTML string to the cypress command log, and having few of this kind of assertions make cypress run out of memory.

**Comment by bahmutov** [comment, marked as answer]
You can skip logging them

```js
cy.request('/').its('body', {log: false})
  .then(text => {
    if (!text.includes(somestring)) {
      throw new Error('Text does not have the expected string')
    }
  })
```

**Reply by stychu** [reply]
mmm. My problem is not with the `.its('body')` but with the `.should('contain', somestring)` or `expect(text).include(somestring)`.
The assertion method Logs the whole `text` to the Command Log and there seems no way to truncate the output.
<img width="367" alt="Screenshot 2020-12-16 at 14 26 37" src="https://user-images.githubusercontent.com/9383657/102474546-10059380-4059-11eb-943c-ffff3a4e38c0.png">

One way around it would be something like
```
cy.request('/').its('body')
  .then(text => {
    expect(text.includes(somestring).true
  })
```

**Reply by bahmutov** [reply]
right - that's why I threw an error instead of using `expect` assertion. 

In more general, check this blog post https://glebbahmutov.com/blog/ssr-e2e/ where I do what I think you would like. For example, you could even load the HTML string into the current document to see it

```js
it('renders 5 items on the server', () => {
  cy.request('/')
    .its('body')
    .then(html => {
      const $li = Cypress.$(html).find('li')
      expect($li)
        .to.have.property('length')
        .equal(4)
      cy.state('document').write(html)
    })
})
```

**Reply by stychu** [reply]
Oh god this is what I was looking for!!!
`cy.state('document').write(html)` (doesn't seem to be documented? can't find any reference in documentation to this)
Thank you!

**Reply by bahmutov** [reply]
Yeah, it is undocumented. I have updated the blog post to use the standard approach
```js
cy.document().invoke({ log: false }, 'write', html)
```

See the repo https://github.com/bahmutov/react-server-example to see it in action

**Reply by stychu** [reply]
Awesome!! Thank you for this article! @bahmutov 
One problem I still have is that when I have multiple tests (describe/it) for use cases the `cy.document().invoke({ log: false }, 'write', html)` appends the iFrame instead of overwriting it which causes assertion errors because the same elements appear multiple times in the iFrame. 
How should I approach clearing the iFrame before writing new HTML to the document?

**Reply by stychu** [reply]
doing `cy.document().invoke('open').invoke({ log: false }, 'write', html)` seems to work.

**Reply by OmarSlame** [reply]
@bahmutov Thanks for this article!
I'm facing a problem in writing the HTML to the document in case if the test already visited a website.
Here is the code snippet:
`cy.visit('any url')`
`// wait until the page is loaded`
`cy.document().invoke({ log: true}, 'write', html)`

Here is what i got:
![image](https://user-images.githubusercontent.com/39995077/200582772-a418c718-df9d-46d1-9dcb-390dabe6dfd8.png)

Can you help me in this?

**Reply by bahmutov** [reply]
@OmarSlame can you provide a small repo with a reproducible example?

**Reply by jprealini** [reply]
@OmarSlame did you find a solution for that issue? I am facing the same exact problem

Thanks!

**Reply by OmarSlame** [reply]
@jprealini Yes, the below working in all cases:
` cy.document().then((document) => {
      document.documentElement.innerHTML = EMAIL_HTML_BODY_YOU_WANT_TO_WRITE
    })`
